test/integration/multi_envoy_test.cc

```diff
@@ -0,0 +1,64 @@
+#include "test/integration/http_protocol_integration.h"
+
+namespace Envoy {
+namespace {
+
+class MultiEnvoyTest : public HttpProtocolIntegrationTest {
+public:
+  // Create an envoy in front of the original Envoy.
+  void createL1Envoy();
+  ~MultiEnvoyTest() { test_server_.reset(); }
+
+  IntegrationTestServerPtr l1_server_;
+  Thread::SkipAsserts skip_;
+};
+
+void MultiEnvoyTest::createL1Envoy() {
+  std::vector<uint32_t> ports;
+  std::vector<uint32_t> zero;
+  int l2_port = lookupPort("http");
+  for (auto& upstream : fake_upstreams_) {
+    if (upstream->localAddress()->ip()) {
+      ports.push_back(l2_port);
+      zero.push_back(0);
+    }
+  }
+  ConfigHelper l1_helper(version_, *api_,
+                         MessageUtil::getJsonStringFromMessageOrDie(config_helper_.bootstrap()));
+  l1_helper.setPorts(zero, true); // Zero out ports set by config_helper_'s finalize();
+  const std::string bootstrap_path = finalizeConfigWithPorts(l1_helper, ports, use_lds_);
+
+  std::vector<std::string> named_ports;
+  const auto& static_resources = config_helper_.bootstrap().static_resources();
+  named_ports.reserve(static_resources.listeners_size());
+  for (int i = 0; i < static_resources.listeners_size(); ++i) {
+    named_ports.push_back(static_resources.listeners(i).name() + "_l1");
+  }
+  createGeneratedApiTestServer(bootstrap_path, named_ports, {false, true, false}, false,
+                               l1_server_);
+}
+
+INSTANTIATE_TEST_SUITE_P(IpVersions, MultiEnvoyTest,
+                         testing::ValuesIn(HttpProtocolIntegrationTest::getProtocolTestParams(
+                             {Http::CodecType::HTTP1}, {Http::CodecType::HTTP1})),
+                         HttpProtocolIntegrationTest::protocolTestParamsToString);
+
+// A simple test to make sure that traffic can flow between client - L1 - L2 - upstream.
+// This test does not currently support mixed protocol hops, or much of the other envoy test
+// framework knobs.
+TEST_P(MultiEnvoyTest, SimpleRequestAndResponse) {
```

Comment:
Nice!

Would it also make sense (perhaps as a follow up?) to have a test in which we create 2 distinct Envoy instances, both in front of the same (or different, I suppose) upstreams. Then we could verify that requests flow through both, close one, and verify that the other still works?

Reply:
yeah I also plan on having tests where I tear down the L2 and verify the L1 500s correctly, or tear down L1 and make sure I can talk directly to L2 -> client, to ensure lifetime issues are really sorted and regression tested against.

Reply:
Having some tests that verify different shutdown ordering would be useful I think, just to make sure that they can exist independently of each other